### PR TITLE
Handle stream io error

### LINKED ISSUE #350: Handle Netstream read/write error
Handle `Stream*` error when `stream.write(…)`/`stream.read(…)`

### PR DESCRIPTION
**Pull request by NIC619**
Fix #350 
For `stream.read`, except for two examples `chat.py` and `echo.py`, it's actually handled in `_handle_ping`.
For `stream.write`, except for two examples `chat.py` and `echo.py`, it's handled in
- ping protocol handler
- identify protocol handler
- `publish` in floodsub and gossipsub
- `_handle_new_peer` and `message_all_peers` in pubsub

### CODE REVIEW

File: libp2p/crypto/rsa.py [hunk 1]
```diff
@@ -25,7 +25,7 @@ def verify(self, data: bytes, signature: bytes) -> bool:
         h = SHA256.new(data)
         try:
             # NOTE: the typing in ``pycryptodome`` is wrong on the arguments to ``verify``.
-            pkcs1_15.new(self.impl).verify(h, signature)  # type: ignore
+            pkcs1_15.new(self.impl).verify(h, signature)
```
Comment (pipermerriam) [inline libp2p/crypto/rsa.py]: looks like the comment above this can be removed tool.

File: libp2p/pubsub/pubsub.py [hunk 2]
```diff
@@ -400,7 +406,13 @@ def handle_subscription(
...
         for stream in self.peers.values():
             # Write message to stream
-            await stream.write(encode_varint_prefixed(raw_msg))
+            try:
+                await stream.write(encode_varint_prefixed(raw_msg))
+            except StreamClosed:
+                peer_id = stream.muxed_conn.peer_id
+                logger.debug("Fail to message peer %s: stream closed", peer_id)
+                del self.peers[peer_id]
+                self.router.remove_peer(peer_id)
```
Comment (mhchia) [inline libp2p/pubsub/pubsub.py]: Should we just use `self._handle_dead_peer(peer_id)` here?

File: libp2p/pubsub/gossipsub.py [hunk 3]
```diff
@@ -596,4 +605,7 @@ def _get_in_topic_gossipsub_peers_from_minus(
         peer_stream = self.pubsub.peers[to_peer]
 
         # Write rpc to stream
-        await peer_stream.write(encode_varint_prefixed(rpc_msg))
+        try:
+            await peer_stream.write(encode_varint_prefixed(rpc_msg))
+        except StreamClosed:
+            logger.debug("Fail to emit control message to %s: stream closed", to_peer)
```
Comment (mhchia) [inline libp2p/pubsub/gossipsub.py]: Should we also do `self._handle_dead_peer(to_peer)` here?

File: libp2p/pubsub/pubsub.py [hunk 4]
```diff
@@ -429,7 +435,12 @@ def handle_subscription(
         # Broadcast message
         for stream in self.peers.values():
             # Write message to stream
-            await stream.write(encode_varint_prefixed(raw_msg))
+            try:
+                await stream.write(encode_varint_prefixed(raw_msg))
+            except StreamClosed:
+                peer_id = stream.muxed_conn.peer_id
+                logger.debug("Fail to message peer %s: stream closed", peer_id)
+                self._handle_dead_peer(peer_id)
```
Comment (mhchia) [inline libp2p/pubsub/pubsub.py]: Just a note: I'm thinking we can somehow extract the `try: ... except StreamClosed: ... self. _handle_dead_peer(peer_id)` to another function in the future.

**Pull request merged**